waspc/src/Analyzer/Decl.hs

```diff
@@ -0,0 +1,36 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TupleSections #-}
+
+module Analyzer.Decl
+  ( Decl (..),
+    takeDecls,
+  )
+where
+
+import Analyzer.TypeDefinitions (IsDeclType)
+import Data.Maybe (mapMaybe)
+import Data.Typeable (Typeable, cast)
+
+-- | Used to store a heterogenous lists of evaluated declarations during
+--   evaluation.
+data Decl where
+  -- | @Decl "Name" value@ results from a declaration statement "type Name value".
+  Decl :: (Typeable a, IsDeclType a) => String -> a -> Decl
+
+-- | Extracts all declarations of a certain type from a @[Decl]@s
+--
+--  Example:
+--
+--  @
+--  data Person = Person { name :: String, age :: Integer } deriving Generic
+--  data Building = Building { address :: String } deriving Generic
+--  let decls = [ Decl "Bob" $ Person "Bob" 42
+--              , Decl "Office" $ Building "1 Road St"
+--              , Decl "Alice" $ Person "Alice" 32
+--              ]
+--  takeDecls @Person decls == [("Bob", Person "Bob" 42), ("Alice", Person "Alice" 32)]
+--  @
+takeDecls :: (Typeable a, IsDeclType a) => [Decl] -> [(String, a)]
```

Comment:
@matijaSos you will find this interesting, with this method we could have avoided some of the boilerplate we had in the old parser. Trick is to use polymorphic return type + `cast`.